src/main/java/com/amazonaws/services/kinesis/clientlibrary/lib/worker/ShardSyncTaskManager.java

```diff
@@ -85,13 +84,20 @@
         this.executorService = executorService;
         this.initialPositionInStream = initialPositionInStream;
         this.shardSyncer = shardSyncer;
+        this.shardSyncRequestPending = new AtomicBoolean(false);
+        this.lock = new ReentrantLock(Boolean.TRUE);
     }
 
-    synchronized Future<TaskResult> syncShardAndLeaseInfo(List<Shard> latestShards) {
-        return checkAndSubmitNextTask(latestShards);
+    Future<TaskResult> syncShardAndLeaseInfo(List<Shard> latestShards) {
+        try {
+            lock.lock();
+            return checkAndSubmitNextTask(latestShards);
+        } finally {
+            lock.unlock();
+        }
     }
 
-    private synchronized Future<TaskResult> checkAndSubmitNextTask(List<Shard> latestShards) {
+    private Future<TaskResult> checkAndSubmitNextTask(List<Shard> latestShards) {
         Future<TaskResult> submittedTaskFuture = null;
```

Comment:
I believe the lock in syncShardAndLeaseInfo should be moved here, so that this method is self contained. The future callers of this method, should not be bothered about acquiring a lock.

Reply:
Tried it. IMO, current implementation looks more readable. I think it's fine as-is, since checkAndSubmitNextTask is a private method.